integration/spark2/src/main/scala/org/apache/spark/sql/execution/strategy/MixedFormatHandler.scala

```diff
@@ -0,0 +1,280 @@
+import org.apache.carbondata.core.datastore.impl.FileFactory
+import org.apache.carbondata.core.metadata.{AbsoluteTableIdentifier, SegmentFileStore}
+import org.apache.carbondata.core.readcommitter.ReadCommittedScope
+import org.apache.carbondata.core.statusmanager.{FileFormat => CarbonFileFormat, SegmentStatus}
+import org.apache.carbondata.core.util.{CarbonProperties, CarbonSessionInfo, SessionParams, ThreadLocalSessionInfo}
+import org.apache.carbondata.core.util.path.CarbonTablePath
+
+object MixedFormatHandler {
+
+  val LOGGER = LogServiceFactory.getLogService(this.getClass.getName)
+
+  val supportedFormats: Seq[String] =
+    Seq("carbon", "carbondata", "parquet", "orc", "json", "csv", "text")
+
+  def validateFormat(format: String): Boolean = {
+    supportedFormats.exists(_.equalsIgnoreCase(format))
+  }
+
+  def getSchema(sparkSession: SparkSession,
+      options: Map[String, String],
+      segPath: String): StructType = {
+    val format = options.getOrElse("format", "carbondata")
+    if ((format.equals("carbondata") || format.equals("carbon"))) {
+      new SparkCarbonFileFormat().inferSchema(sparkSession, options, Seq.empty).get
+    } else {
+      val filePath = FileFactory.addSchemeIfNotExists(segPath.replace("\\", "/"))
+      val path = new Path(filePath)
+      val fs = path.getFileSystem(SparkSQLUtil.sessionState(sparkSession).newHadoopConf())
+      val status = fs.listStatus(path, new PathFilter {
+        override def accept(path: Path): Boolean = {
+          !path.getName.equals("_SUCCESS") && !path.getName.endsWith(".crc")
+        }
+      })
+      getFileFormat(new CarbonFileFormat(format)).inferSchema(sparkSession, options, status).get
+    }
+  }
+
+  def extraRDD(l: LogicalRelation,
+      projects: Seq[NamedExpression],
+      filters: Seq[Expression],
+      readCommittedScope: ReadCommittedScope,
+      identier: AbsoluteTableIdentifier,
+      supportBatch: Boolean = true): Option[(RDD[InternalRow], Boolean)] = {
+    val loadMetadataDetails = readCommittedScope.getSegmentList
+    val segsToAccess = getSegmentsToAccess(identier)
+    val rdds = loadMetadataDetails.filterNot(l =>
+      l.getFileFormat.equals(CarbonFileFormat.COLUMNAR_V3) ||
+      l.getFileFormat.equals(CarbonFileFormat.ROW_V1) &&
+      (!(l.getSegmentStatus.equals(SegmentStatus.SUCCESS) &&
+         l.getSegmentStatus.equals(SegmentStatus.LOAD_PARTIAL_SUCCESS))))
+      .filter(l => segsToAccess.isEmpty || segsToAccess.contains(l.getLoadName))
+      .groupBy(_.getFileFormat)
+      .map { case (format, detailses) =>
+        val paths = detailses.flatMap { d =>
+          SegmentFileStore.readSegmentFile(CarbonTablePath.getSegmentFilePath(readCommittedScope
+            .getFilePath, d.getSegmentFile)).getLocationMap.asScala.flatMap { case (p, f) =>
+            f.getFiles.asScala.map { ef =>
+              p + CarbonCommonConstants.FILE_SEPARATOR + ef
+            }.toSeq
+          }.toSeq
+        }.map(new Path(_))
```

Comment:
instead of this map we can return new Path(p + CarbonCommonConstants.FILE_SEPARATOR + ef) on line 100

Reply:
ok